### PR TITLE
OpenStack Validations and Doc Updates for Floating IPs

### PR DESCRIPTION
**Pull request by iamemilio**
The user interface for floating IPs has been changed such that
the ingress and api floating ips can no longer be passed when
the external network is unset. These validations check that this
usage is respected, while also checking the the floating IPs passed
by the user are valid and available.

/label platform/openstack
/cc @Fedosin @adduarte

### CODE REVIEW

File: docs/user/openstack/customization.md [hunk 1]
```diff
@@ -23,6 +23,7 @@ Beyond the [platform-agnostic `install-config.yaml` properties](../customization
 * `externalDNS` (optional list of strings): The IP addresses of DNS servers to be used for the DNS resolution of all instances in the cluster
 * `externalNetwork` (optional string): Name of external network the installer will use to provide access to the cluster. If defined, a floating ip from this network will be created and associated with the bootstrap node to facilitate debugging and connection to the bootstrap node during installation. The lbFloatingIP property is a floating ip address selected from this network.
 * `lbFloatingIP` (optional string): Address of existing Floating IP from externalNetwork the installer will associate with the API load balancer. This property is only valid if externalNetwork is defined. If externalNetwork is not defined, this property is ignored.
```
Comment (mandre) [inline docs/user/openstack/customization.md:25]: We actually error out now when the externalNetwork is not defined. Let's update the doc.

File: docs/user/openstack/customization.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ Beyond the [platform-agnostic `install-config.yaml` properties](../customization
...
+* `ingressFloatingIP` (optional string): Address of an existing Floating IP from externalNetwork the installer will associate with the ingress port. This property is only valid if externalNetwork is defined. If externalNetwork is not defined, this property is ignored.
```
Comment (mandre) [inline docs/user/openstack/customization.md:26]: ditto

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go [hunk 3]
```diff
@@ -123,3 +131,34 @@ func (ci *CloudInfo) getNetwork(networkName string) (*networks.Network, error) {
...
+
+func (ci *CloudInfo) getFloatingIPs(api string, ingress string) (API *floatingips.FloatingIP, Ingress *floatingips.FloatingIP, err error) {
+	API = &floatingips.FloatingIP{}
+	Ingress = &floatingips.FloatingIP{}
+
+	if api != "" {
+		fip, err := floatingips.Get(ci.clients.networkClient, api).Extract()
+		if err != nil {
+			var gerr gophercloud.ErrDefault404
+			if !errors.As(err, &gerr) {
```
Comment (mandre) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:143]: This comparison is wrong, it should be:
```suggestion
			if errors.As(err, &gerr) {
```

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go [hunk 4]
```diff
@@ -123,3 +131,34 @@ func (ci *CloudInfo) getNetwork(networkName string) (*networks.Network, error) {
...
+			return nil, nil, err
+		}
+		API = fip
+	}
+
+	if ingress != "" {
+		fip, err := floatingips.Get(ci.clients.networkClient, ingress).Extract()
+		if err != nil {
+			var gerr gophercloud.ErrDefault404
+			if !errors.As(err, &gerr) {
```
Comment (mandre) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:155]: Same here:
```suggestion
			if errors.As(err, &gerr) {
```

**Comment by mandre** [comment]
Seems like the unit tests do not cover the case where gophercloud returns a 404 :)

**Comment by iamemilio** [comment]
> Seems like the unit tests do not cover the case where gophercloud returns a 404 :)

That is handled with this case:
```go
if ci.APIFIP == nil {
    allErrs = append(allErrs, field.NotFound(fldPath.Child("lbFloatingIP"), p.LbFloatingIP))
}
```

returning nil means that it was not found

**Comment by iamemilio** [comment]
> Seems like the unit tests do not cover the case where gophercloud returns a 404 :)

https://github.com/openshift/installer/blob/637b2bfd4d7a6a8a095d6573cd4486496b417950/pkg/asset/installconfig/openstack/validation/platform_test.go#L77-L112

**Comment by mandre** [comment]
> > Seems like the unit tests do not cover the case where gophercloud returns a 404 :)
> 
> https://github.com/openshift/installer/blob/637b2bfd4d7a6a8a095d6573cd4486496b417950/pkg/asset/installconfig/openstack/validation/platform_test.go#L77-L112

I was talking about ignoring 404 errors after the resource was deleted by the installer. This would have caught the previous error with errors.As() comparison.

**Comment by iamemilio** [comment]
> > > Seems like the unit tests do not cover the case where gophercloud returns a 404 :)
> > 
> > 
> > https://github.com/openshift/installer/blob/637b2bfd4d7a6a8a095d6573cd4486496b417950/pkg/asset/installconfig/openstack/validation/platform_test.go#L77-L112
> 
> I was talking about ignoring 404 errors after the resource was deleted by the installer. This would have caught the previous error with errors.As() comparison.

This conversation was taken offline, however the conclusion was that unit tests for the cloudinfo package are needed

**Comment by mandre** [comment]
/lgtm
/approve

CI is not cooperating at the moment, but it at least shows that it has passed validation. We should be good to go.

**Pull request merged**